Initialization DB. Part Mikhail D.

grey3228 opened the pull request:
Сделал все свои таблички. Сразу скажу, что insert'ы в таблицу delivery_sessions можно было придумать и интереснее (на данный момент, в каждой сессии только 1 заказ, т.е. курьеры будто после каждого выполненного заказа прекращали сессию, а потом стартовали новую), но у меня в заказах даты заданы не статически, а определяются динамически (через CURRENT_DATE и т.д.). Поэтому, чтобы долго не возиться, сделал пока так. Но в дальнейшем нам ничто не помешает руками в случае необходимости несколько таких сессий накидать.
Конфиг правил так в соответствии с тем, что мы договорились перейти на postgis. @swenkal, если у тебя такой вариант не заработает, то будем менять, у меня всё ок.

grey3228 committed:
Adding my part of sql scripts and correcting .properties

grey3228 committed:
Changing README.md

grey3228 committed:
Small scheme fixes due to previous commits

grey3228 committed:
Fixing some mistakes in schema.sql

swenkal commented on README.md:
````diff
@@ -3,10 +3,10 @@
+- run postgis image:
 ```bash
-docker run --name pg-delivery -p 5432:5432 -e POSTGRES_PASSWORD=postgres -d postgres
+docker run --name pg-delivery -p 5432:5432 -e POSTGRES_PASSWORD=postgis -e POSTGRES_USER=postgis -d postgis/postgis
````
В целом все равно, но все же интересно. Чем тебя имя БД, пользователя и пароля помешали то? :D

grey3228 replied:
Ахахах, да ничем, мне это ни в коем случае не принципиально) Просто когда менял readme, нужно было менять название образа с dockerhub (postgis/postgis), вот я заодно чёт всё остальное тоже поменял. Не вопрос, откачу имена бд и юзера, пароль обратно, чтобы привычнее выглядело

swenkal commented on src/main/resources/initDB/schema.sql:
```diff
@@ -1,5 +1,125 @@
+            DECLARE pos BIGINT;
+            BEGIN
+                FOR pos IN (SELECT product_position_id FROM (SELECT * FROM orders INNER JOIN orders_product_positions USING(order_id) WHERE order_id = NEW.order_id) positions) LOOP
+                UPDATE product_positions orders_product_positions SET current_amount = current_amount + (SELECT amount FROM orders INNER JOIN orders_product_positions USING(order_id) WHERE order_id = NEW.order_id AND product_position_id = pos) WHERE product_position_id = pos; 
```
Тут случайно нет ошибки? может нужно убрать **orders_product_positions** в начале ?  
```suggestion
                UPDATE product_positions SET current_amount = current_amount + (SELECT amount FROM orders INNER JOIN orders_product_positions USING(order_id) WHERE order_id = NEW.order_id AND product_position_id = pos) WHERE product_position_id = pos; 
```

swenkal commented on src/main/resources/initDB/schema.sql:
```diff
@@ -1,5 +1,125 @@
+
+
+CREATE TABLE IF NOT EXISTS product_positions(product_position_id BIGSERIAL PRIMARY KEY,
+                                             product_id BIGINT NOT NULL,
```
Не увидел FOREIGN KEY для **product_id**

swenkal commented on src/main/resources/application.properties:
```diff
@@ -1,9 +1,11 @@
+spring.jpa.properties.hibernate.dialect = org.hibernate.spatial.dialect.postgis.PostgisDialect
+spring.sql.init.mode=always
+spring.sql.init.encoding= UTF-8
+spring.sql.init.schema-locations=classpath:/initDB/schema.sql
+spring.sql.init.data-locations=classpath:/initDB/data.sql
```
Опытным путем пришел к таким пропертям для инициализации. Пока больше ни разу не валились у меня. Предлагаю их использовать. 
```suggestion
spring.datasource.schema=classpath:initDB/schema.sql
spring.datasource.data=classpath:initDB/data.sql
```

grey3228 replied:
Такие свойства помечаются Deprecated. Мб всё же не использовать?

swenkal reviewed (changes requested):
Есть парочку моментов для уточнения, также есть предложения по исправлениям в properties и в одной из функций в schema.sql

grey3228 committed:
Update src/main/resources/initDB/schema.sql

Вроде как всё корректно, т.к. по правилам после UPDATE нужно указывать все таблицы, которые в этом запросе далее будут использоваться. Я по началу тоже не добавлял orders_product_positions, но бд заругалась

Co-authored-by: Александр <41340273+swenkal@users.noreply.github.com>

grey3228 committed:
Cancelling changes from last wrong commit

grey3228 commented:
Вместо оставления коммента по поводу комментария @swenkal (см. https://github.com/netcracker-edu/2021-msk-food-delivery/pull/5/commits/6a607ac357de8ad6399f0c4d8ba20d7c0f407331) прожал на принятие изменений :/

grey3228 committed:
Fixing scripts, properties, readme

swenkal commented on src/main/resources/application.properties:
```diff
@@ -1,9 +1,14 @@
+spring.sql.init.schema-locations=classpath:initDB/schema.sql
+spring.sql.init.data-locations=classpath:initDB/data.sql
+
+spring.jackson.deserialization.FAIL_ON_UNKNOWN_PROPERTIES=true
```
Затесалось из реализации уже product_positions!
в принципе, предлагаю оставить. Ничего криминального, потом все равно добавлять.

swenkal reviewed (approved):
Одна помарочка, но незначительная абсолютно.
Уже давно пора мерджить  данный PR и пробовать запуститься со всеми данными Миши и моими.

iLeonidze reviewed (approved):
Нужно поправить конфликты и последние вопросы от Александра, в остальном ок

grey3228 committed:
Final merged scripts

grey3228 committed:
Correcting typo in .properties

grey3228 committed:
Merge branch 'main' into grey_db_init_scripts

grey3228 committed:
Adding dependency for postgis geom types
